compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -387,8 +391,91 @@ fn link_rlib<'a, B: ArchiveBuilder<'a>>(
+    //
+    // Note, though, that we don't actually want this metadata to show up in any
+    // final output of the compiler. Instead this is purely for rustc's own
+    // metadata tracking purposes.
+    //
+    // With the above in mind, each "flavor" of object format gets special
+    // handling here depending on the target:
+    //
+    // * MachO - macos-like targets will insert the metadata into a section that
+    //   is sort of fake dwarf debug info. Inspecting the source of the macos
+    //   linker this causes these sections to be skipped automatically because
+    //   it's not in an allowlist of otherwise well known dwarf section names to
+    //   go into the final artifact.
+    //
+    // * WebAssembly - we actually don't have any container format for this
+    //   target. WebAssembly doesn't support the `dylib` crate type anyway so
+    //   there's no need for us to support this at this time. Consequently the
+    //   metadata bytes are simply stored as-is into an rlib.
+    //
+    // * COFF - Windows-like targets create an object with a section that has
+    //   the `IMAGE_SCN_LNK_REMOVE` flag set which ensures that if the linker
+    //   ever sees the section it doesn't process it and it's removed.
+    //
+    // * ELF - All other targets are similar to Windows in that there's a
+    //   `SHF_EXCLUDE` flag we can set on sections in an object file to get
+    //   automatically removed from the final output.
+    //
+    // Note that this metdata format is kept in sync with
+    // `rustc_codegen_llvm/src/metadata.rs`.
+    fn create_metadata_object(sess: &Session, metadata: &[u8]) -> Vec<u8> {
+        let endianness = match sess.target.options.endian {
+            Endian::Little => Endianness::Little,
+            Endian::Big => Endianness::Big,
+        };
+        let architecture = match &sess.target.arch[..] {
+            "arm" => Architecture::Arm,
+            "aarch64" => Architecture::Aarch64,
+            "x86" => Architecture::I386,
+            "s390x" => Architecture::S390x,
+            "mips" => Architecture::Mips,
+            "x86_64" => Architecture::X86_64,
+
+            // This is used to handle all "other" targets. This includes targets
+            // like WebAssembly and those otherwise just not having support here
+            // at this time. For these targets the metadata file in the rlib is
+            // simply the raw metadata itself. This would break dylib
+            // compliation if it were supported, but for most of these targets
+            // dylib compilation is not supported anyway (such as WebAssembly).
+            _ => return metadata.to_vec(),
+        };
+
+        if sess.target.is_like_osx {
+            let mut file = Object::new(BinaryFormat::MachO, architecture, endianness);
+
+            let section =
+                file.add_section(b"__DWARF".to_vec(), b".rustc".to_vec(), SectionKind::Debug);
```

Comment:
Coming back to this, I'm somewhat worried about reusing the debug segments as a place to store our metadata. Debug info sections seems like something that toolchains may want to inspect in their own ways. AFAICT there are many kinds that are likely better suited here: `Other`, `Metadata`, `Note` etc, that are less likely to be of interest to anybody, and that don't interfere with purpose `__DWARF` or similar segments.

Reply:
I can see that `Mach-O` might not have anything else suitable for our purposes, in which case its fine to use `Debug`… I guess. But concern still extends equally to other targets.

Reply:
Notes end up in the linked executable and I think only exist in ELF. For writing there are only three standard segments: text, data and debug.

https://github.com/gimli-rs/object/blob/2e0af3f6867db3482da40bd8ab1989319da56e78/src/write/macho.rs#L38

On the reading side sections are never classified as other, metadata or note by the macho reader.

https://github.com/gimli-rs/object/blob/2e0af3f6867db3482da40bd8ab1989319da56e78/src/read/macho/section.rs#L202

Reply:
Is it necessary to use the `__DWARF` segment for this section? Would it work if we named it something else?

Reply:
Maybe?

Reply:
I determined the `__DWARF` here by poking around in the [source of ld64 itself](https://opensource.apple.com/source/ld64/ld64-609/src/ld/parsers). I came upon [this file](https://opensource.apple.com/source/ld64/ld64-609/src/ld/parsers/macho_relocatable_file.cpp.auto.html) where if you search for `__DWARF` in that file you'll see a loop in `Parser<A>::makeSections()`. The goal was to make this loop drop the section with metadata, effectively triggering a `continue` somewhere.

I actually previously looked at this souce when LLVM bitcode was merged into object files since this confirmed that on macOS the bitcode setions were dropped from the final output. Rummaging around this loop the best way that seemed to trick the linker to not include the section in the final output was a `__DWARF` section with an unknown name, so I settled on that. The `SectionKind::Debug` and `__DWARF` name were both chosen to ensure that the branch there in the loop was hit.